### PR TITLE
[tune] Auto-init Ray + default SearchAlg

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?
If `ray.init` hasn't been called, Tune should automatically handle that.

It'd be nice not to need to import ray and call ray.init in the general
case; it's not a necessary concept to get started with Tune.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/tune/tune.py [hunk 1]
```diff
@@ -90,7 +90,8 @@ def run(run_or_experiment,
         queue_trials=False,
         reuse_actors=False,
         trial_executor=None,
-        raise_on_failed_trial=True):
+        raise_on_failed_trial=True,
+        ray_redis_address="auto"):
```
Comment (ericl) [inline python/ray/tune/tune.py]: Do we need to handle redis address too? It's probably easier to only handle the local init case, which can be done without passing in any extra args in `tune.run` only.

File: python/ray/tune/tests/test_trial_runner.py [hunk 2]
```diff
@@ -2070,7 +2070,7 @@ def testCheckpointWithFunction(self):
             checkpoint_freq=1)
         tmpdir = tempfile.mkdtemp()
         runner = TrialRunner(
-            BasicVariantGenerator(), metadata_checkpoint_dir=tmpdir)
+            , metadata_checkpoint_dir=tmpdir)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
            metadata_checkpoint_dir=tmpdir)
```

File: python/ray/tune/tests/test_trial_runner.py [hunk 3]
```diff
@@ -2092,7 +2092,7 @@ def count_checkpoints(cdir):
         trial = Trial("__fake", checkpoint_freq=1)
         tmpdir = tempfile.mkdtemp()
         runner = TrialRunner(
-            BasicVariantGenerator(), metadata_checkpoint_dir=tmpdir)
+            , metadata_checkpoint_dir=tmpdir)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
            metadata_checkpoint_dir=tmpdir)
```

File: python/ray/tune/tests/test_trial_runner.py [hunk 4]
```diff
@@ -1954,7 +1954,7 @@ def testTrialSaveRestore(self):
         tmpdir = tempfile.mkdtemp()
 
         runner = TrialRunner(
-            BasicVariantGenerator(), metadata_checkpoint_dir=tmpdir)
+            , metadata_checkpoint_dir=tmpdir)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
            metadata_checkpoint_dir=tmpdir)
```

File: python/ray/tune/tests/test_trial_runner.py [hunk 5]
```diff
@@ -2014,7 +2014,7 @@ def testTrialNoSave(self):
...
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
            metadata_checkpoint_dir=tmpdir)
```

File: python/ray/tune/tests/test_trial_runner.py [hunk 6]
```diff
@@ -977,7 +977,7 @@ def tearDown(self):
         _register_all()  # re-register the evicted objects
 
     def generate_trials(self, spec, name):
-        suggester = BasicVariantGenerator()
+        suggester =
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
        suggester = BasicVariantGenerator()
```

File: python/ray/tune/tests/test_trial_runner.py [hunk 7]
```diff
@@ -1209,7 +1209,7 @@ def train(config, reporter):
         }
 
         for name, spec in experiments.items():
-            trial_generator = BasicVariantGenerator()
+            trial_generator =
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_runner.py]:
```suggestion
            trial_generator = BasicVariantGenerator()
```

File: python/ray/tune/ray_trial_executor.py [hunk 8]
```diff
@@ -33,12 +33,17 @@ def unwrap(self):
         return self._result
 
 
+def _auto_init_ray():
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
```

File: python/ray/tune/ray_trial_executor.py [hunk 9]
```diff
@@ -33,12 +33,17 @@ def unwrap(self):
...
+    pass
```
Comment (richardliaw) [inline python/ray/tune/ray_trial_executor.py]:
```suggestion
```

File: python/ray/tune/tune.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/tune.py]: Done

File: python/ray/tune/ray_trial_executor.py [hunk 10]
```diff
@@ -33,12 +33,15 @@ def unwrap(self):
         return self._result
 
 
+
+
```
Comment (robertnishihara) [inline python/ray/tune/ray_trial_executor.py]: This whitespace must have been accidental.

File: python/ray/tune/tune.py [hunk 11]
```diff
@@ -90,7 +90,8 @@ def run(run_or_experiment,
         queue_trials=False,
         reuse_actors=False,
         trial_executor=None,
-        raise_on_failed_trial=True):
+        raise_on_failed_trial=True,
+        ray_auto_init=True):
```
Comment (robertnishihara) [inline python/ray/tune/tune.py]: update the docstring

**Review by ericl — APPROVED** [review]

**Pull request merged**